tests/decomposition/test_controlled_decomposition.py

```diff
@@ -0,0 +1,599 @@
+                    },
+                ): 1,
+            }
+        )
+
+
+@pytest.mark.parametrize(
+    "op, decomp, resources, custom_op_type",
+    [
+        # Single-qubit controlled on 0
+        (
+            qml.ops.Controlled(qml.Z(1), control_wires=[0], control_values=[0]),
+            [qml.X(0), qml.CZ(wires=[0, 1]), qml.X(0)],
+            to_resources({qml.CZ: 1, qml.X: 2}),
+            qml.ops.CZ,
+        ),
+        # Single-qubit controlled on 1
+        (
+            qml.ops.Controlled(qml.Z(1), control_wires=[0], control_values=[1]),
+            [qml.CZ(wires=[0, 1])],
+            to_resources({qml.CZ: 1}),
+            qml.ops.CZ,
+        ),
+        # Two-qubit controlled
+        (
+            qml.ops.Controlled(qml.Z(2), control_wires=[0, 1], control_values=[1, 0]),
+            [qml.X(1), qml.CCZ(wires=[0, 1, 2]), qml.X(1)],
+            to_resources({qml.CCZ: 1, qml.X: 2}),
+            qml.ops.CCZ,
+        ),
+        # Parametrized controlled
+        (
+            qml.ops.Controlled(qml.RX(0.5, wires=1), control_wires=[0], control_values=[0]),
+            [qml.X(0), qml.CRX(0.5, wires=[0, 1]), qml.X(0)],
+            to_resources({qml.CRX: 1, qml.X: 2}),
+            qml.ops.CRX,
+        ),
+        # Controlled on two qubits
+        (
+            qml.ops.Controlled(qml.SWAP(wires=[1, 2]), control_wires=[0], control_values=[0]),
+            [qml.X(0), qml.CSWAP(wires=[0, 1, 2]), qml.X(0)],
+            to_resources({qml.CSWAP: 1, qml.X: 2}),
+            qml.CSWAP,
+        ),
+    ],
+)
+def test_controlled_to_custom_controlled(op, decomp, resources, custom_op_type):
+    """Tests that a controlled op decomposes to its corresponding custom op if applicable."""
+
+    decomp_rule = CustomControlledDecomposition(custom_op_type)
+    with qml.queuing.AnnotatedQueue() as q:
+        decomp_rule(*op.parameters, wires=op.wires, **op.hyperparameters)
+
+    assert q.queue == decomp
+    assert decomp_rule.compute_resources(**op.resource_params) == resources
+
+
+class CustomOp(qml.operation.Operation):  # pylint: disable=too-few-public-methods
+    """A custom op."""
+
+    resource_param_keys = ("num_wires",)
+
+    @property
+    def resource_params(self):
+        return {"num_wires": len(self.wires)}
+
+
+def _custom_resource(num_wires):
+    return {
+        qml.X: 1,
+        qml.CNOT: 1,
+        qml.Toffoli: 1,
+        qml.resource_rep(
+            qml.MultiControlledX,
+            num_control_wires=3,
+            num_zero_control_values=1,
+            num_work_wires=1,
+        ): 1,
+        qml.RX: 1,
+        qml.Rot: 1,
+        qml.CRZ: 1,
+        resource_rep(qml.MultiRZ, num_wires=num_wires): 1,
+        controlled_resource_rep(
+            qml.MultiRZ,
+            {"num_wires": num_wires - 1},
+            num_control_wires=1,
+        ): 1,
+        resource_rep(qml.PauliRot, pauli_word="XYX"): 1,
+        qml.Z: 1,
+        qml.CZ: 1,
+    }
```

Comment:
Is there a specific motivation behind this decomposition, or is it just "some blob"? :)

Reply:
It's just "some blob" that has a bit of everything.